mypyc/build.py

```diff
@@ -0,0 +1,373 @@
+                           cfile: str,
+                           build_dir: str,
+                           extra_compile_args: List[str],
+                           ) -> List[MypycifyExtension]:
+    """Produce the list of extension modules when a shared library is needed.
+
+    This creates one shared library extension module that all of the
+    others link against and then one shim extension module for each
+    module in the build, that simply calls an initialization function
+    in the shared library.
+
+    We treat the shared library as a python extension so that it is
+    cleanly processed by setuptools, but it isn't *really* a python C
+    extension module on its own.
+    """
+    shared_lib = MypycifyExtension(
+        'lib' + lib_name,
+        is_mypyc_shared=True,
+        sources=[cfile],
+        include_dirs=[include_dir()],
+        extra_compile_args=extra_compile_args,
+    )
+    extensions = [shared_lib]
+
+    for source in sources:
+        module_name = source.module.split('.')[-1]
+        stub_file = generate_c_extension_shim(source.module, module_name, build_dir)
+
+        # We include the __init__ in the "module name" we stick in the Extension,
+        # since this seems to be needed for it to end up in the right place.
+        full_module_name = source.module
+        assert source.path
+        if os.path.split(source.path)[1] == '__init__.py':
+            full_module_name += '.__init__'
+        extensions.append(MypycifyExtension(
+            full_module_name,
+            mypyc_shared_target=shared_lib,
+            sources=[stub_file],
+            extra_compile_args=extra_compile_args,
+        ))
+
+    return extensions
+
+
+def build_single_module(sources: List[BuildSource],
+                        cfile: str,
+                        extra_compile_args: List[str],
+                        ) -> List[MypycifyExtension]:
+    """Produce the list of extension modules for a standalone extension.
+
+    This contains just one module, since there is no need for a shared module.
+    """
+    return [MypycifyExtension(
+        sources[0].module,
+        sources=[cfile],
+        include_dirs=[include_dir()],
+        extra_compile_args=extra_compile_args,
+    )]
+
+
+def mypycify(paths: List[str],
+             mypy_options: Optional[List[str]] = None,
+             opt_level: str = '',
+             skip_cgen: bool = False) -> List[MypycifyExtension]:
+    """Main entry point to building using mypyc.
+
+    This produces a list of Extension objects that should be passed as the
+    ext_modules parameter to setup.
+
+    Arguments:
+      * paths: A list of file paths to build. It may contain globs.
+      * mypy_options: Optionally, a list of command line flags to pass to mypy.
+                      (This can also contain additional files, for compatability reasons.)
+      * opt_level: The optimization level, as a string. Defaults to '' (meaning '-O').
```

Comment:
For clang, '-O' is equivalent to `-O2', but maybe you are right and we should default to 3?

Clang does not because 3 takes longer and generates larger code and historically doesn't always make things better. In our testing it has, though, so...